### PR TITLE
Fixed replacement of outputs

### LINKED ISSUE #515: ComputationGraph.replace doesn't work on graph outputs
The following example fails:
```python
>>> from theano import tensor
>>> from blocks.graph import ComputationGraph
>>> x = tensor.scalar()
>>> y = x + 1
>>> cg = ComputationGraph([y])
>>> doubled_cg = cg.replace([(y, 2 * y)])
>>> doubled_cg.outputs[0].eval({x: 2})
3.0
```

### PR DESCRIPTION
**Pull request by dmitriy-serdyuk**
Fixes #515.

### CODE REVIEW

File: blocks/graph.py [hunk 1]
```diff
@@ -209,6 +209,13 @@ def replace(self, replacements):
                 replacement_keys_cur.append(input_)
                 replacement_vals_cur.append(replacements[input_])
 
+        # Add outputs of the computation graph
+        for output in self.outputs:
+            if output not in replacements:
```
Comment (vdumoulin) [inline blocks/graph.py]: The following would be more straightforward:
```python
if output in replacements:
    replacement_keys_cur.append(output)
    replacement_vals_cur.append(replacements[output])
```

Comment (dmitriy-serdyuk) [inline blocks/graph.py]: Is it a bad practice to use `continue`? I tried to decrease the number of indents.

Comment (bartvm) [inline blocks/graph.py]: I don't think `continue` is discouraged per se, and it can often be the right choice. In this case I agree with @vdumoulin though; `if .. in ..` just reads a bit more natural.

**Comment by dwf** [comment]
LGTM.

File: blocks/graph.py [hunk 2]
```diff
@@ -204,10 +204,15 @@ def replace(self, replacements):
         # Sort `replacements` in topological order
         for node in apply_nodes_sorted:
             for input_ in node.inputs:
-                if input_ not in replacements:
-                    continue
-                replacement_keys_cur.append(input_)
-                replacement_vals_cur.append(replacements[input_])
+                if input_ in replacements:
```
Comment (rizar) [inline blocks/graph.py]: There is a bug here, because a variable can be an input of multiple nodes.

Comment (vdumoulin) [inline blocks/graph.py]: Can you give an example of where this would fail?

Comment (dwf) [inline blocks/graph.py]: A failing test case, even, would be good.

Comment (rizar) [inline blocks/graph.py]: Sure, I will. I got it when if was monitoring mean activation of a dropped-out layer, that is the values from this layers were used for monitoring and cost computations.

It took me a while to find this bug and this is the main reason why I am lagging behind in a few important discussions here.

Comment (vdumoulin) [inline blocks/graph.py]: I get the cause, but what impact does this bug have on the correctness of computation?

Comment (rizar) [inline blocks/graph.py]: It made me panic for two days, because after I started to use the aforementioned monitoring applying dropout _caused_  severe overfitting. This sort of makes sense, because division by `1 - dropout_prob` happened twice and the training and validation objective were no longer similar, but quite different.

I will fix it tomorrow, if nobody does it before me.

Comment (vdumoulin) [inline blocks/graph.py]: OH! I get it, I think. So you're saying that if a node to be replaced is an input of multiple nodes in the graph, its replacement will happen more than once which will affect the correctness of the computation.

Here's a minimal example that reproduces the issue:
```python
from theano import tensor
from theano.printing import debugprint
from blocks.graph import ComputationGraph

x = tensor.scalar('x')
y = 2 * x
z = x + 1

cg = ComputationGraph([y, z])
debugprint(cg.outputs[0])
cg = cg.replace({x: 0.5 * x})
print('----------')
debugprint(cg.outputs[0])
print('----------')
print cg.outputs[0].eval({x: 1.0}), cg.outputs[1].eval({x: 1.0})
```

The expected output for `y` and `z` is 1 and 1.5, respectively, but the output produced by the code snippet above is 0.5 and 1.25, which corresponds to the replacement being applied twice on `x`, as shown in the `debugprint` output.

@rizar I'll make a pull request for that right away.

Comment (rizar) [inline blocks/graph.py]: Exactly! Sorry for not explaining it as clearly as you did.

> @rizar I'll make a pull request for that right away.

Cool, thanks!

Comment (rizar) [inline blocks/graph.py]: It seems to me that the whole iteration could be actually replaced by `for variable in self.variables`. They are also sorted topologically, exactly in the way as done here (correct me if I am wrong), but the duplicates are removed, which is just what we need here.

Comment (vdumoulin) [inline blocks/graph.py]: @rizar see my PR in #525

Comment (vdumoulin) [inline blocks/graph.py]: I think you're right, `self.variables` is topologically sorted.

Comment (vdumoulin) [inline blocks/graph.py]: Although it also contains auxiliary variables. Is it an issue?

Comment (rizar) [inline blocks/graph.py]: Needs a bit of thought. So far it is definitely not because replacements of auxiliary variables is not supported, see #514. If we support it later we will have to be able to traverse all variables including auxiliaries in topological order, but again, the right tool for it would be `self.variables`.

So I would say it's not, maybe we should mention it in a comment in `replace`.
